README.md
# intmed
The intmed is for conducting mediation analysis using the interventional effect approach. It is built upon the work by VanderWeele, Vansteelandt and Robins (2014; https://dx.doi.org/10.1097/ede.0000000000000034) , Vansteelandt and Daniel (2017; https://dx.doi.org/10.1097/ede.0000000000000596) and Chan and Leung (2020; unpublished manuscript, available on request). The indirect effect mediated through a mediator M is defined as the average difference in the potential outcome in the population when there is an intervention that shifts the distribution of the mediator from what would have realised from unexposed to exposed, while holding the exposure level constant as exposed. 


NEWS.md
# Update 27.8.2020
- This update fixes (1) a bug when the user requested multiple imputation, and (2) presentation of the results when multiple imputation is used.
- The update also allows the input data to be multiply imputed dataset in long format.


tests/testthat.R
library(testthat)
library(intmed)

test_check("intmed")


tests/testthat/test-mediate.R
context("Testing the mediate function")

set.seed(1234)



###################################################
# sim_data example - 1 mediator
###################################################

data("sim_data", package = "intmed")
tol = 0.006
tol2 = 0.009
med_res <- mediate(y = "y", med = c("m"), treat = "x", ymodel = "regression", mmodel = c("regression"), treat_lv = 1, control_lv = 0, incint = TRUE, inc_mmint = FALSE, data = sim_data, sim = 5000, digits = 5, HTML_report = FALSE, cores = 2)

test_that("Test using sim_data - 1 mediator", {
  skip_on_cran()
  skip_on_travis()
  expect_equal(mean(med_res$combined$direct), 0.523, tol)
  expect_equal(mean(med_res$combined$indirect[[1]]), 0.207, tol)
  expect_equal(mean(med_res$combined$total), 0.730, tol)
  expect_equal(median(med_res$combined$prop[[1]]), 0.283, tol2)
})


###################################################
# substance example - 2 mediators
###################################################
#tol = 0.006
#tol2 = 0.0015

#data("substance", package = "intmed")
#med_res <- intmed::mediate(y = "sub_misuse", med = c("dev_peer","sub_exp"), treat = "fam_int", c = c("conflict","gender"), ymodel = "logistic regression", mmodel = c("logistic regression","logistic regression"), treat_lv = 1, control_lv = 0, data = substance, sim = 1000, digits = 5, HTML_report = FALSE, cores = 2)

#test_that("Test using substance data - 2 mediators", {
#  skip_on_cran()
#  skip_on_travis()
#  expect_equal(mean(med_res$combined$direct), -0.05318 , tol)
#  expect_equal(mean(med_res$combined$indirect[[1]]), -0.01811, tol)
#  expect_equal(mean(med_res$combined$indirect[[2]]), -0.00622, tol)
#  expect_equal(mean(med_res$combined$total), -0.07670, tol)
#  expect_equal(median(med_res$combined$prop[[1]]), 0.22564, tol2)
#  expect_equal(median(med_res$combined$prop[[2]]), 0.07174, tol2)
#})

###################################################
# simulating data - 3 mediator
###################################################

#tol = 0.1

#n = 10000
#x = runif(n)
#x = ifelse(x < 0.5, 0, 1)

#c = rnorm(n,0,0.1)

#m1 = 0.2*x + rnorm(n, 0, 0.1) + 0.1*c
#m2 = 0.3*x + 0.9*m1 + rnorm(n, 0, 0.1) + 0.2*c
#m3 = 0.4*x + m1 + 1.1*m2 + rnorm(n, 0, 0.1) + 0.3*c

#y = 0.8*x + 0.5*m1 + 0.6*m2 + 0.7*m3 + rnorm(n, 0, 0.1) + 2 + 0.4*c

#simdata <- data.frame(x,m1,m2,m3,y,c)
#med_res <- mediate(y = "y", med = c("m1","m2","m3"), treat = "x", c = "c", ymodel = "regression", mmodel = c("regression", "regression", "regression"), treat_lv = 1, control_lv = 0, incint = NULL, inc_mmint = FALSE, data = simdata, sim = 5000, digits = 5, cores = 2)

#test_that("Test using simulated data - 3 mediators", {
#  skip_on_cran()
#  skip_on_travis()
#  expect_equal(mean(med_res$combined$direct), 0.8, tol)
#  expect_equal(mean(med_res$combined$indirect[[1]]), 0.1, tol)
#  expect_equal(mean(med_res$combined$indirect[[2]]), 0.288, tol)
#  expect_equal(mean(med_res$combined$indirect[[3]]), 0.7896, tol)
#  expect_equal(mean(med_res$combined$total), 1.9776, tol)
#})


###################################################
# simulating data with missing data - 3 mediators
###################################################

#simdata$rand = runif(nrow(simdata),0,1)
#simdata$y[simdata$rand < 0.1] <- NA
#simdata$rand = runif(nrow(simdata),0,1)
#simdata$x[simdata$rand < 0.05] <- NA
#simdata$rand = runif(nrow(simdata),0,1)
#simdata$m1[simdata$rand < 0.02] <- NA
#simdata$rand = runif(nrow(simdata),0,1)
#simdata$m2[simdata$rand < 0.01] <- NA
#simdata$rand = runif(nrow(simdata),0,1)
#simdata$m3[simdata$rand < 0.01] <- NA

#med_res <- mediate(y = "y", med = c("m1","m2","m3"), treat = "x", c = "c", ymodel = "regression", mmodel = c("regression", "regression", "regression"), treat_lv = 1, control_lv = 0, incint = NULL, inc_mmint = FALSE, data = simdata, sim = 1000, digits = 5)

#test_that("Test using simulated data with missing data - 3 mediators", {
#  skip_on_cran()
#  expect_equal(mean(med_res$combined$direct), 0.8, tol)
#  expect_equal(mean(med_res$combined$indirect[[1]]), 0.1, tol)
#  expect_equal(mean(med_res$combined$indirect[[2]]), 0.288, tol)
#  expect_equal(mean(med_res$combined$indirect[[3]]), 0.7896, tol)
#  expect_equal(mean(med_res$combined$total), 1.9776, tol)
#})


R/impute.R
mi_prepare_impute <- function(fo, data, maxit = 20) {
  df <- extract_analysis_vars(data, fo)
  fo_terms <- stats::terms(fo)
  fo_term_labels <- base::attr(fo_terms,"term.labels")
  fo_vars <- base::all.vars(fo)
  fo_length <- length(fo_vars)
  fo_interactions <- stringr::str_subset(fo_term_labels, ":")

  mi_formulas <- list()
  DV <- ""
  for (i in seq_along(fo_vars)) {
    DV <- fo_vars[i]
    search_term <- stringr::str_c("(^",DV,":)|(:",DV,":)|(:", DV, "$)|(^",DV,"$)")
    IVs <- fo_term_labels[!stringr::str_detect(fo_term_labels, search_term)]
    IVs <- paste0(IVs, collapse = " + ")
    if (i > 1) {
      IVs <- paste0(fo_vars[1], "+",  IVs)
    }
    tmp_formula <- stats::as.formula(paste0(DV, " ~ " ,IVs))
    mi_formulas <- c(mi_formulas, tmp_formula)
  }

  mi_formulas <- mice::name.formulas(mi_formulas)
  max_missing_perc <- sum(ifelse(rowSums(as.data.frame(lapply(df, is.na))) == 0, 0 ,1))/nrow(df)*100

  if (max_missing_perc > 0 & max_missing_perc < 5) {
    max_missing_perc = 5
  }else {
    max_missing_perc = floor(max_missing_perc)
  }

  prepare_impute_obj <- list(DV = fo_vars[[1]], IVs = fo_term_labels, formulas = mi_formulas,
                             data = df, m = max_missing_perc, maxit = maxit)

  class(prepare_impute_obj) <- "prepare_impute_obj"
  return(prepare_impute_obj)
}




R/utility_function.R
char2fac <- function(data, ...) {
  df <- dplyr::as_tibble(data)
  df[, !sapply(df, is.numeric) & !sapply(df, is.factor)] <- lapply(df[, !sapply(df, is.numeric) & !sapply(df, is.factor)], as.factor)
  return(df)
}

extract_analysis_vars <- function(data, fo, imputed_data = FALSE, ...) {
  df <- dplyr::as_tibble(data)
  fo_vars <- base::all.vars(fo)
  if (imputed_data == TRUE) {
    df <- df[, c(fo_vars, ".imp", ".id")]
  }else {
    df <- df[, fo_vars]
  }
  return(df)
}

empirical_pvalue <- function(vec) {
  gtzero <- ifelse(vec > 0, 1, 0)
  p <- sum(gtzero)/length(gtzero)
  p <- ifelse(p > 0.5, 1-p, p)
  p <- 2*p
  return(p)
}

validate_input <- function(y, med, treat, mod, c, ymodel, mmodel, treat_lv, control_lv, incint, inc_mmint, data, sim, conf.level, out_scale, complete_analysis, digits) {
  tmp <- NULL
  search_term = paste0("^",y,"$")
  if (sum(stringr::str_detect(names(data), search_term)) != 1) {
    stop("The outcome variable does not exist in the dataset. Check the variable name.")
    return(FALSE)
  }


  for (i in 1:length(med)) {
    search_term = paste0("^",med[i],"$")
    if (sum(stringr::str_detect(names(data), search_term)) != 1) {
      stop(paste0("Mediator ", i," does not exist in the dataset. Check the variable name."))
    }
  }

  search_term = paste0("^",treat,"$")
  if (sum(stringr::str_detect(names(data), search_term)) != 1) {
    stop("The treatment variable does not exist in the dataset. Check the variable name.")
    return(FALSE)
  }

  if (!is.null(c)) {
    for (i in 1:length(c)) {
      search_term = paste0("^",c[i],"$")
      if (sum(stringr::str_detect(names(data), search_term)) != 1) {
        stop(paste0("Covariate ", i, " does not exist in the dataset. Check the variable name."))
      }
    }
  }

  search_term = paste0("^",ymodel,"$")
  if (sum(stringr::str_detect(c("regression","poisson regression","logistic regression"),search_term)) != 1) {
    stop("Only regression, poisson regerssion and logistic are supported for the outcome variable.")
  }

  if (ymodel == "logistic regression") {

    expr = parse(text = paste0("tmp <- is.factor(data$",y,")"))
    eval(expr)
    if (tmp) {
      expr = parse(text = paste0("tmp <- length(levels(data$",y,"))"))
      eval(expr)
      if (tmp != 2) {
        stop("Logistic regression is specified for the outcome variable. The outcome must have 2 levels. Please check the variable.")
      }
    }else {
      stop("Logistic regression is specified for the outcome variable. The outcome must be a factor (categorical) variable. Please check the variable and convert it into factor using as.factor().")
    }
  }

  if (ymodel == "poisson regression") {
    expr = parse(text = paste0("tmp <- is.numeric(data$",y,")"))
    eval(expr)
    if (!tmp) {
      stop("Poisson regression is specified for the outcome variable. The outcome must be a numeric variable. Please check the variable.")
    }else {
      expr = parse(text = paste0("tmp <- sum(data$",y," %% 1 != 0, na.rm = TRUE)"))
      eval(expr)
      if (tmp != 0) {
        stop("Poisson regression is specified for the outcome variable. The outcome must be integers. Please check the variable.")
      }else {
        expr = parse(text = paste0("tmp <- sum(data$",y," < 0, na.rm = TRUE)"))
        eval(expr)
        if (tmp != 0) {
          stop("Poisson regression is specified for the outcome variable. The outcome must be non-negative. Please check the variable.")
        }
      }
    }
  }

  if (ymodel == "regression") {
    expr = parse(text = paste0("tmp <- is.numeric(data$",y,")"))
    eval(expr)
    if (!tmp) {
      stop("Regression is specified for the outcome variable. The outcome must be a numeric variable. Please check the variable.")
    }
  }

  #Need to test the following 10.1.2020
  if (length(med) != length(mmodel)) {
    stop("The number of mediators must be the same as the number of models specified for the mediators.")
  }

  for (i in 1:length(med)) {
    search_term = paste0("^",mmodel[i],"$")
    if (sum(stringr::str_detect(c("regression","poisson regression","logistic regression"),search_term)) != 1) {
      stop("Only regression, poisson regerssion and logistic are supported for the mediator(s).")
    }

    if (mmodel[i] == "logistic regression") {
      expr = parse(text = paste0("tmp <- is.factor(data$",med[i],")"))
      eval(expr)
      if (tmp) {
        expr = parse(text = paste0("tmp <- length(levels(data$",med[i],"))"))
        eval(expr)
        if (tmp != 2) {
          stop(paste0("Logistic regression is specified for mediator ", i,". This mediator must have 2 levels. Please check the variable."))
        }
      }else {
        stop(paste0("Logistic regression is specified for mediator ",i,". This mediator must be a factor (categorical) variable. Please check the variable and convert it into factor using as.factor()."))
      }
    }

    if (mmodel[i] == "poisson regression") {
      expr = parse(text = paste0("tmp <- is.numeric(data$",med[i],")"))
      eval(expr)
      if (!tmp) {
        stop("Poisson regression is specified for mediator ",i,". This mediator must be a numeric variable. Please check the variable.")
      }else {
        expr = parse(text = paste0("tmp <- sum(data$",med[i]," %% 1 != 0, na.rm = TRUE)"))
        eval(expr)
        if (tmp != 0) {
          stop("Poisson regression is specified for mediator ",i,". This mediator must be integers. Please check the variable.")
        }else {
          expr = parse(text = paste0("tmp <- sum(data$",med[i]," < 0, na.rm = TRUE)"))
          eval(expr)
          if (tmp != 0) {
            stop("Poisson regression is specified for mediator ",i,". The outcome must be non-negative. Please check the variable.")
          }
        }
      }
    }

    if (mmodel[i] == "regression") {
      expr = parse(text = paste0("tmp <- is.numeric(data$",med[i],")"))
      eval(expr)
      if (!tmp) {
        stop("Regression is specified for mediator ",i,". This mediator must be a numeric variable. Please check the variable.")
      }
    }
  }

}


R/gen_html_results.R
gen_med_reg_html <- function(res_df, y, med, treat, c, ymodel, mmodel, incint = NULL, inc_mmint = FALSE, conf.level = 0.95, data_head, treat_lv = NULL, control_lv = NULL) {
  html_output = "<br/><h4><u>Regression analysis</u></h4>The table below shows the estimates from the key regression models for the mediation analysis."
  html_line = paste0("<ul><li>The outcome variable, ",y,", was modelled with ", ymodel,"</li>")
  html_output = paste0(html_output, html_line)
  for (i in 1:length(med)) {
    html_line = paste0("<li>The mediator, ", med[i], ", was modelled with ", mmodel[i], "</li>")
    html_output = paste0(html_output, html_line)
  }
  html_output = paste0(html_output, "</ul>")

  alpha_lv = 1-conf.level
  coeff_lab = rep("b", length(med)+1)

  html_line = paste0("The mediator ", ifelse(length(med) > 1, "models ", "model ") ,"indicated that the exposure variable, ", treat,", was ")
  html_output = paste0(html_output, html_line)
  ##Stop here - need to take care of binary mediator and binary treatment 7.1.2020##
  expr = parse(text = paste0("tmp = is.factor(data_head$",treat,")"))
  eval(expr)
  treat_row_name = ifelse(tmp,paste0(treat,treat_lv), treat)
  html_line = "<ul>"
  for (i in 1:length(med)) {
    html_line = paste0(html_line, "<li>")
    tmp = paste0(ifelse(res_df[res_df$variables == treat_row_name,3*i+1] < alpha_lv, "significantly ","not significantly "), "associated with the mediator, ", med[[i]],", ")
    html_line = paste0(html_line, tmp)

    if (mmodel[[i]] == "logistic regression") {
      coeff_lab[i] = "OR"
    }else if (mmodel[[i]] == "poisson regression") {
      coeff_lab[[i]] = "IRR"
    }

    tmp = paste0(coeff_lab[i], " = ", stringr::str_replace_all(res_df[res_df$variables == treat_row_name, 3*(i-1)+2],"\\*",""),", ", round(conf.level*100, 0), "% CI = ", res_df[res_df$variables == treat_row_name, 3*(i-1)+3],"." )

    html_line = paste0(html_line, tmp, "</li>")
  }
  html_line = paste0(html_line, "</ul>")

  html_output = paste0(html_output, html_line)

  if (ymodel == "logistic regression") {
    coeff_lab[length(med)+1] = "OR"
  }else if (ymodel == "poisson regression") {
    coeff_lab[length(med)+1] = "IRR"
  }

  if (is.null(incint) | sum(incint) == 0) {
    html_line = paste0("After accounting for the effect of the ", ifelse(length(med) > 1, "mediators ", "mediator "), ", the outcome model indicated that the treatment variable, ", treat, ", was ")
    html_line = paste0(html_line, ifelse(res_df[res_df$variables == treat_row_name,3*length(med)+4] < alpha_lv, "significantly ","not significantly "),"associated with ", y, ", ")
    html_line = paste0(html_line, coeff_lab[length(med)+1], " = ", stringr::str_replace_all(res_df[res_df$variables == treat_row_name, 3*length(med)+2],"\\*",""),", ", round(conf.level*100, 0), "% CI = ", res_df[res_df$variables == treat_row_name, 3*length(med)+3],"." )
    html_output = paste0(html_output, html_line, "<br/>")
    if (inc_mmint == FALSE) {
      html_line = "The outcome model also indicated that <ul>"
      for (i in 1:length(med)) {
        expr = parse(text = paste0("tmp = is.factor(data_head$",med[i],")"))
        eval(expr)
        med_row_name <- NULL
        expr = parse(text = paste0("med_row_name = ifelse(tmp, paste0(med[i],tail(levels(data_head$",med[i],"),1)),med[i])"))
        eval(expr)
        html_line = paste0(html_line, "<li>",med[i]," is ", ifelse(res_df[res_df$variables == med_row_name,3*length(med)+4] < alpha_lv, "significantly ","not significantly "), "associated with the outcome, ", y, ", ")
        html_line = paste0(html_line, coeff_lab[length(med)+1], " = ", stringr::str_replace_all(res_df[res_df$variables == med_row_name,3*length(med)+2],"\\*",""), ", ", round(conf.level*100,0),"% CI = ", res_df[res_df$variables == med_row_name,3*length(med)+3],". </li>")
      }
      html_line = paste0(html_line, "</ul>")
      html_output = paste0(html_output, html_line)
    }else {
      html_line = "The outcome model includes the mediator-mediator interaction(s). The effect of the mediators and their interaction(s) are summarized in the three right-most columns in the table below. <br/>"
      html_output = paste0(html_output, html_line)
    }
  }else {
    if (inc_mmint == TRUE) {
      html_line = "The outcome model includes the treatment-mediator interaction(s) and also the mediator-mediator interaction(s). The effect of the treatment, mediator(s) and relevant interactions are summarized in the three right-most columns in the table below. <br/>"
      html_output = paste0(html_output, html_line)
    }else {
      html_line = "The outcome model includes the treatment-mediator interaction(s). The effect of the treatment, mediator(s) and relevant interaction(s) are summarized in the three right-most columns in the table below. <br/>"
      html_output = paste0(html_output, html_line)
    }
  }

  if (!is.null(c)) {
    if (length(c) == 1) {
      html_line = paste0("The effect of ", c," was adjusted for in the above analyses. <br/>")
    }else {
      tmp = paste(paste0(c[1:length(c)-1], collapse = ", "), c[length(c)], sep = " and ")
      html_line = paste0("The effect of ", tmp," were adjusted for in the above analyses. <br/>")
    }
    html_output = paste0(html_output, html_line)
  }

  html_output = c(html_output,"<br/><b>Table. Results from key regression analyses.</b><br/>")
  html_output = c(html_output, "<table style = \"text-align: left;border-bottom: 1px solid black; border-top: 1px solid black;\" cellspacing=\"0\" cellpadding = \"2\">")
  html_line = "<tr><td></td>"

  for (i in 1:length(med)) {
    html_line = paste0(html_line,"<td colspan = \"3\">",med[i],"</td>")
  }

  html_line = paste0(html_line, "<td colspan = \"3\">", y,"</td></tr>")
  html_output = c(html_output, html_line)
  html_line = "<tr><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">Variables</td>"

  tmp = paste0(paste0("<td style=\"padding-right: 1em;border-bottom: 1px solid black;\">",coeff_lab,"</td><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">", round(conf.level*100,0),"% CI","</td><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">p-value</td>"), collapse = "")

  html_line = paste0(html_line, tmp, "</tr>")
  html_output = c(html_output, html_line)

  res_df[sapply(res_df, is.na)] = "-"
  for (i in 1:nrow(res_df)) {
    html_line = paste0("<tr><td style=\"padding-right: 1em\">",paste0(res_df[i,], collapse = "</td><td style=\"padding-right: 1em\">"),"</td></tr>")
    html_output = c(html_output, html_line)
  }

  html_output = c(html_output, "</table>")
  html_output = c(html_output, "***<i>p</i> <.001;**<i>p</i> <.01;*<i>p</i> <.05. <br/>")
  html_output = c(html_output, "<br/>Reference<br/>Rubin DB. Multiple imputation for nonresponse in surveys. New York: John Wiley & Sons; 2009.")
  html_output = c(html_output, "<br/>Buuren Sv, Groothuis-Oudshoorn K. mice: Multivariate imputation by chained equations in R. Journal of statistical software. 2010:1-68.")
  html_output = c(html_output, "<br/>Vansteelandt S, Daniel RM. Interventional effects for mediation analysis with multiple mediators. Epidemiology (Cambridge, Mass). 2017; 28(2):258.")
  html_output = c(html_output, "<br/>Chan GCK, Leung J. Causal mediation analysis using the interventional effect approach. A refined definition and software implementation in R. Paper under review. 2020.")

  return(html_output)
}

gen_med_table_html <- function(med_res, med, conf.level = 0.95, digits = 2, sim) {
  res_des = paste0("<br/><h4><u>Mediation analysis</u></h4>Mediation analysis was performed based on the counter-factual framework and the interventional effect (Vansteelandt and Daniel, 2017; Chan and Leung, 2020). The analysis was conducted in R using the intmed package (Chan and Leung, 2020) with ",sim, " simulations. <br/> Results from the mediation analysis indicated that <ul>")
  alpha_level = 1-conf.level
  html_output = "<br/><b>Table. Mediation analysis results.</b><br/>"
  html_output = c(html_output, "<table style = \"text-align: left;border-bottom: 1px solid black; border-top: 1px solid black;\" cellspacing=\"0\" cellpadding = \"2\">")
  html_line = paste0("<tr><td style=\"padding-right: 1em;border-bottom: 1px solid black;\"></td><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">Estimates</td><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">",round(conf.level*100, 0),"% CI</td><td style=\"padding-right: 1em;border-bottom: 1px solid black;\">p-value</td></tr>")
  html_output = c(html_output, html_line)

  for (i in 1:length(med_res$indirect)) {
    estimate <- mean(med_res$indirect[[i]])
    ci <- stats::quantile(med_res$indirect[[i]], c((1-conf.level)/2,1-(1-conf.level)/2))
    pvalue <- empirical_pvalue(med_res$indirect[[i]])
    formatted_p <- format(round(pvalue,3), nsmall = 3)
    html_line <- paste0("<tr><td style=\"padding-right: 1em\">Indirect effect mediated through ", med[i],"</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),ifelse(pvalue < 0.05,"*",""),ifelse(pvalue < 0.01,"*",""),ifelse(pvalue < 0.001,"*",""),"</td><td style=\"padding-right: 1em\">(",format(round(ci[1], digits), nsmall = digits),", ",format(round(ci[2], digits), nsmall = digits),")</td><td>",formatted_p,"</td></tr>")
    html_output = c(html_output, html_line)
    res_des_line = paste0("<li> the indirect effect through ",med[i], " was ", ifelse(pvalue < alpha_level, "statistically significant","non-signifcant"),", p = ",formatted_p,"; ")
    res_des_line = paste0(res_des_line, format(round(stats::median(med_res$prop[[i]])*100, digits = digits),nsmall = digits), "% of the total effect was mediated through ", med[i], ".</li>")
    res_des = c(res_des, res_des_line)

  }

  if (!is.null(med_res$dependence)) {
    estimate <- mean(med_res$dependence)
    ci <- stats::quantile(med_res$dependence, c((1-conf.level)/2,1-(1-conf.level)/2))
    pvalue <- empirical_pvalue(med_res$dependence)
    formatted_p <- format(round(pvalue,3), nsmall = 3)
    html_line <- paste0("<tr><td style=\"padding-right: 1em\">Indirect effect mediated through the dependence between mediators</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),ifelse(pvalue < 0.05,"*",""),ifelse(pvalue < 0.01,"*",""),ifelse(pvalue < 0.001,"*",""),"</td><td style=\"padding-right: 1em\">(",format(round(ci[1], digits), nsmall = digits),", ",format(round(ci[2], digits), nsmall = digits),")</td><td>",formatted_p,"</td></tr>")
    html_output = c(html_output, html_line)
    res_des = c(res_des, paste0("<li> the indirect effect through the dependence between mediators was ", ifelse(pvalue < alpha_level, "statistically significant","non-signifcant"),", p = ",formatted_p,".</li>"))
  }

  if (!is.null(med_res$interaction)) {
    estimate <- mean(med_res$interaction)
    ci <- stats::quantile(med_res$interaction, c((1-conf.level)/2,1-(1-conf.level)/2))
    pvalue <- empirical_pvalue(med_res$interaction)
    formatted_p <- format(round(pvalue,3), nsmall = 3)
    html_line <- paste0("<tr><td style=\"padding-right: 1em\">Indirect effect mediated through the interaction between mediators</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),ifelse(pvalue < 0.05,"*",""),ifelse(pvalue < 0.01,"*",""),ifelse(pvalue < 0.001,"*",""),"</td><td style=\"padding-right: 1em\">(",format(round(ci[1], digits), nsmall = digits),", ",format(round(ci[2], digits), nsmall = digits),")</td><td>",formatted_p,"</td></tr>")
    html_output = c(html_output, html_line)
    res_des = c(res_des, paste0("<li> the indirect effect through the interaction between mediators was ", ifelse(pvalue < alpha_level, "statistically significant","non-signifcant"),", p = ",formatted_p,".</li>"))
  }

  estimate <- mean(med_res$direct)
  ci <- stats::quantile(med_res$direct, c((1-conf.level)/2,1-(1-conf.level)/2))
  pvalue <- empirical_pvalue(med_res$direct)
  formatted_p <- format(round(pvalue,3), nsmall = 3)
  html_line <- paste0("<tr><td style=\"padding-right: 1em\">Direct effect</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),ifelse(pvalue < 0.05,"*",""),ifelse(pvalue < 0.01,"*",""),ifelse(pvalue < 0.001,"*",""),"</td><td style=\"padding-right: 1em\">(",format(round(ci[1], digits), nsmall = digits),", ",format(round(ci[2], digits), nsmall = digits),")</td><td>",formatted_p,"</td></tr>")
  html_output = c(html_output, html_line)
  res_des = c(res_des, paste0("<li> the direct effect was ", ifelse(pvalue < alpha_level, "statistically significant","non-signifcant"),", p = ",formatted_p,".</li>"))

  estimate <- mean(med_res$total)
  ci <- stats::quantile(med_res$total, c((1-conf.level)/2,1-(1-conf.level)/2))
  pvalue <- empirical_pvalue(med_res$total)
  formatted_p <- format(round(pvalue,3), nsmall = 3)
  html_line <- paste0("<tr><td style=\"padding-right: 1em\">Total effect</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),ifelse(pvalue < 0.05,"*",""),ifelse(pvalue < 0.01,"*",""),ifelse(pvalue < 0.001,"*",""),"</td><td style=\"padding-right: 1em\">(",format(round(ci[1], digits), nsmall = digits),", ",format(round(ci[2], digits), nsmall = digits),")</td><td>",formatted_p,"</td></tr>")
  html_output = c(html_output, html_line)
  res_des = c(res_des, paste0("<li> the total effect was ", ifelse(pvalue < alpha_level, "statistically significant","non-signifcant"),", p = ",formatted_p,".</li>"))

  for (i in 1:length(med_res$prop)) {
    estimate <- stats::median(med_res$prop[[i]])
    ci <- stats::quantile(med_res$prop[[i]], c((1-conf.level)/2,1-(1-conf.level)/2))
    pvalue <- "-"
    html_line <- paste0("<tr><td style=\"padding-right: 1em\">Proportion of effect mediated through ", med[i],"</td><td style=\"padding-right: 1em\">", format(round(estimate, digits), nsmall = digits),"</td><td style=\"padding-right: 1em\"></td><td></td></tr>")
    html_output = c(html_output, html_line)
  }
  html_output <- c(html_output, "</table>")
  html_output <- c(html_output, "***<i>p</i> < .001;**<i>p</i> < .05***<i>p</i> < .05.")
  res_des <- c(res_des, "</ul>")
  html_output <- c(res_des, html_output)

  return(html_output)
}

gen_med_reg_table <- function(y_res, m_res, med, ymodel, mmodel, conf.level = 0.95, digits = 2) {

  table <- extract_reg_table(m_res[[1]], mmodel[[1]], conf.level, digits = digits)
  colnames(table) <- c("variables", "m1.b","m1.ci","m1.p")

  if (length(m_res) > 1) {
    for (i in 2:length(m_res)) {
      tmp <- extract_reg_table(m_res[[i]], mmodel[[i]], conf.level, digits =  digits)
      colnames(tmp) <- c("variables", paste0("m",i,".b"), paste0("m",i,".ci"), paste0("m",i,".p"))
      table <- merge(table, tmp, by = "variables", all = TRUE)
    }
  }

  tmp <- extract_reg_table(y_res, ymodel, conf.level, digits)

  colnames(tmp) <- c("variables","y.b","y.ci", "y.p")

  table <- merge(table, tmp , by = "variables", all = TRUE)
  table.part1 <- table[!is.na(table$m1.b),]
  table.part2 <- table[is.na(table$m1.b),]
  table <- rbind(table.part1,table.part2)

  return(table)

}

extract_reg_table <- function(res, model, conf.level = 0.95, digits = 2) {
  half_alpha = (1-conf.level)/2
  if (class(res)[[1]] == "mipo") {
    table <- summary(res)
    table$ts <- stats::qt(half_alpha, table$df)
    table$variables <- table$term
  }else if (class(res)[[1]] == "glm" | class(res)[[1]] == "lm"){
    table <- as.data.frame(summary(res)$coefficients)
    table$df <- res$df.residual
    colnames(table)[which(colnames(table) == "Estimate")] = "estimate"
    colnames(table)[which(colnames(table) == "Std. Error")] = "std.error"
    if (class(res)[[1]] == "lm") {
      colnames(table)[which(colnames(table) == "Pr(>|t|)")] = "p.value"
      table$ts <- stats::qt(half_alpha, table$df)
    }else if (class(res)[[1]] == "glm") {
      colnames(table)[which(colnames(table) == "Pr(>|z|)")] = "p.value"
      table$ts <- stats::qnorm(half_alpha)
    }
    table$variables <- rownames(table)
  }


  table$lb <- table$estimate + table$ts*table$std.error
  table$ub <- table$estimate - table$ts*table$std.error

  if (model == "logistic regression" | model == "poisson regression") {
    table$lb = exp(table$lb)
    table$ub = exp(table$ub)
    table$estimate = exp(table$estimate)
  }

  table$ci <- paste0("(",format(round(table$lb, digits), nsmall = digits),", ",format(round(table$ub, digits), nsmall = digits),")")
  table$estimate <- format(round(table$estimate, digits), digits)
  table$estimate <- ifelse(table$p.value < .05 ,paste0(table$estimate,"*"),table$estimate)
  table$estimate <- ifelse(table$p.value < .01 ,paste0(table$estimate,"*"),table$estimate)
  table$estimate <- ifelse(table$p.value < .001 ,paste0(table$estimate,"*"),table$estimate)


  table$p.value <- format(round(table$p.value, 3), nsmall = 3)
  table <- table[,c("variables","estimate","ci","p.value")]
  return(table)
}


